api/src/main/java/org/pmiops/workbench/actionaudit/ActionAuditQueryServiceImpl.java

```diff
@@ -38,11 +38,13 @@
           + "  jsonPayload.prev_value AS prev_value,\n"
           + "  jsonPayload.new_value AS new_value\n"
           + "FROM %s\n"
-          + "WHERE %s AND\n"
-          + "  @after <= TIMESTAMP_MILLIS(CAST(jsonPayload.timestamp AS INT64)) AND\n"
-          + "  TIMESTAMP_MILLIS(CAST(jsonPayload.timestamp AS INT64)) < @before\n"
+          + "WHERE %s\n"
+          + "  AND @after <= TIMESTAMP_MILLIS(CAST(jsonPayload.timestamp AS INT64))\n"
+          + "  AND TIMESTAMP_MILLIS(CAST(jsonPayload.timestamp AS INT64)) < @before\n"
+          + "  AND @after <= _PARTITIONTIME\n"
```

Comment:
Will `@after` always be equal to a day boundary?  If not, this query will miss 23:59:59 worth of events in the worst case.  Choosing `(@after - 1 day <= partition)` is safer, and only scans 1 extraneous partition in the worst case.

Actually... are we guaranteed to have a partition daily?  Maybe increase that safety buffer.

The `@before` check looks fine.

Reply:
You're right. I missed the boundary condition there.

Reply:
Additionally, the event timestamps predate (hopefully!) the ingestion timestamps (which are the basis for partition determination) by delays in logging and sinking to BQ. Maybe a buffer of 1 day on either side is prudent.